API Doc for RadioButtons (#7)
Microsoft/microsoft-ui-xaml-specs: active/radiobuttons/RadioButtonsAPI.md

# Background

> The RadioButtons control is our answer to enabling better keyboarding and narrator support for RadioButtons that need to be presented in a related list.

Reviewer: ```suggestion
The RadioButtons control is our answer to enabling better keyboarding and narrator support for RadioButton elements that need to be presented in a related list.
```

Currently, the only way to group RadioButtons is through using a StackPanel and individually labeling each RadioButton through its GroupName property in order for narrator to read out the options somewhat understandably.

However, using that method still left gaps in the narrator experience and StackPanel does not behave the way a user would expect when navigating the list via keyboard - RadioButtons was introduced to help solve those issues.

# Description

> RadioButtons is a new control that enables you to create related lists of RadioButtons easily, while also fully supporting keyboarding and narrator functionality.

Reviewer: I think it is confusing that RadioButtons is control, but we are refering to collections of objects of the RadioButton class as RadioButtons in this spec as well. Perhaps we could be a little more verbose to avoid this confusion by replacing the RadioButtons which do not refer to the new control with "Groups of RadioButton objects",

> RadioButtons is a new control that enables you to create related lists of RadioButtons easily, while also fully supporting keyboarding and narrator functionality.

Reviewer: I think it is confusing that RadioButtons is a control, but we are referring to collections of objects of the RadioButton class as RadioButtons in this spec as well. Perhaps we could be a little more verbose to avoid this confusion by replacing the RadioButtons which do not refer to the new control with something like "RadioButton Elements".

Author: I see two comments of your that look very similar - are you suggesting either using "RadioButton Elements" or "Groups of RadioButton objects"? Either way I think I understand what you're getting at and will try and re-word it to be less confusing.

Reviewer: Sorry, I thought I deleted the other one. Personally I think RadioButton Elements is cleaner.  The main point is that because of the name of this control I don't think we should use RadioButtons as the plural of RadioButton, since we are trying to name this new control RadioButtons.

Author: Makes sense! I just went through and updated the wording in those sections - take a look and let me know if that helps make it clearer.

Reviewer: Just one more, made a separate comment with the suggested change.

Author: Noted! I committed them.

> # Examples

Author: This unfortunately doesn't really work all that well. We get this:
![image](https://user-images.githubusercontent.com/6964369/53511762-fbb46900-3a75-11e9-9698-451738470e09.png)

@teaP - this seems like a bug, is it known?

> # Examples

Reviewer 2: It's not a bug, you just can't do this.

> # Examples

Reviewer 3: Does the scenario make sense?  It's what apps have to do today.

> # Examples

Author: It's apparently not a scenario we can target at this time. Removing references to in in API doc.

The following example demonstrates how to create a small collective group of RadioButton elements using the RadioButtons control.

```xaml
<RadioButtons Header="App Mode" SelectedIndex="2">
    <x:String>Dark</x:String>
    <x:String>Light</x:String>
    <x:String>Windows Default</x:String>         
</RadioButtons>
```

### DataBinding with RadioButtons
An example of simple data binding of the ItemsSource with RadioButtons.

```xaml
<preview:RadioButtons x:Name="RadioButtonGroup" Header="App Mode" ItemsSource="{x:Bind radioButtonItems}" />
```

```C#
public sealed partial class MainPage : Page
{
    public class OptionDataModel
    {
        public string label;
        public override string ToString()
        {
            return label;
        }
    }

    List<OptionDataModel> radioButtonItems;

    public MainPage()
    {
        this.InitializeComponent();

        radioButtonItems = new List<OptionDataModel>();
        radioButtonItems.Add(new OptionDataModel() { label = "Item 1" });
        radioButtonItems.Add(new OptionDataModel() { label = "Item 2" });
        radioButtonItems.Add(new OptionDataModel() { label = "Item 3" });
    }
}
```

> ### Multiple Columns

Author: The reason for this specific layout when setting multiple columns was to imitate the common and known behavior keyboard users recognized from Win32. @chigy did some extensive research here about why the layout should follow this.

We were also running into consistency issues with our app team partners about spacing and column formatting when they were using StackPanels + RadioButtons in their Settings pages. That's the main reason we decided to leave out too much customization, to keep a pattern with this control that is consistent for everyone.

Some groups of RadioButton elements may want a multi-column layout. This is an example on how to set that up.

```xaml

> <preview:RadioButtons x:Name="RadioButtonGroup" Header="App Mode" MaximumColumns="3">

Author: @teaP Something you can look at as well?

    <x:String>Column 1</x:String>
    <x:String>Column 2</x:String>
    <x:String>Column 3</x:String>
    <x:String>Column 1</x:String>
    <x:String>Column 2</x:String>
    <x:String>Column 3</x:String>
</preview:RadioButtons>
```

![alt text](multicolumns.png)

# API Details

### RadioButtons

```
namespace Microsoft.UI.Xaml.Controls
{
[webhosthidden]
[contentproperty("Items")]
unsealed runtimeclass RadioButtons : Windows.UI.Xaml.Controls.Control
{
    RadioButtons();

    Object ItemsSource;
    Windows.Foundation.Collections.IVector<Object> Items{ get; };
    Windows.UI.Xaml.DataTemplate ItemTemplate;

>     Windows.UI.Xaml.DependencyObject ContainerFromItem(Object item);

Author: @teaP - to confirm but I thought the container type was RadioButtonListViewItem?

    Windows.UI.Xaml.DependencyObject ContainerFromIndex(Int32 index);

    Int32 SelectedIndex;
    Object SelectedItem;
    event Windows.UI.Xaml.Controls.SelectionChangedEventHandler SelectionChanged;

    Int32 MaximumColumns;
    Object Header;

    static Windows.UI.Xaml.DependencyProperty ItemsSourceProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ItemsProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ItemTemplateProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty SelectedIndexProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty SelectedItemProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaximumColumnsProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty HeaderProperty{ get; };
}
}
```

### RadioButtons Automation Pier

```
namespace Microsoft.UI.Xaml.Automation.Peers
{
[webhosthidden]
unsealed runtimeclass RadioButtonsListViewItemAutomationPeer : Windows.UI.Xaml.Automation.Peers.ListViewItemAutomationPeer
{
    RadioButtonsListViewItemAutomationPeer(RadioButtonsListViewItem owner);
}
}
```

### RadioButtons Primitives

```
namespace Microsoft.UI.Xaml.Controls.Primitives
{
[webhosthidden]

> unsealed runtimeclass RadioButtonsListView : Windows.UI.Xaml.Controls.ListView

Author: Not sure exactly what you're referring to... the webhosthidden? @teaP to help.

> unsealed runtimeclass RadioButtonsListView : Windows.UI.Xaml.Controls.ListView

Reviewer 3: No, why is there a RadioButtonsListView and RadioButtonsAutomationPeer?  I think we should add a brief description of these in a "Remarks" section, after the Examples section.

> unsealed runtimeclass RadioButtonsListView : Windows.UI.Xaml.Controls.ListView

Reviewer 3: No, I mean we're not giving any explanation for the RibbonsListView API, or the Ribbon...Peer APIs.  I know they're mostly implementation details, I would just describe them in a "Remarks" section after the Examples section.

{
    RadioButtonsListView();
}

[webhosthidden]
[default_interface]
unsealed runtimeclass RadioButtonsListViewItem : Windows.UI.Xaml.Controls.ListViewItem
{
    RadioButtonsListViewItem();
}
}
```

# API Notes

| Name | Description |
|:-:|:--|

> | Header | Places a text label above the group of RadioButton elements and is read out by the narrator on focus. |

Author: I made a change to this, please check for accuracy.

| HeaderTemplate | Identifies the Header dependency property. |
| SelectedItem | Gets or sets the index of the selected item. Selection is denoted by that item's RadioButton icon being checked. |

> | SelectedIndex | Gets or sets the selected item.|

Author: Was calling out that the selection behavior is slightly different from the selected item behavior you get in a listview - are you suggestion this is too redundant and I should remove?

> | SelectedIndex | Gets or sets the selected item.|

Reviewer 4: Are the descriptions for SelectedItem and SelectedIndex backwards?

> | SelectedIndex | Gets or sets the selected item.|

Author: Good catch! Look to be.

> | SelectedIndex | Gets or sets the selected item.|

Reviewer 3: I was just saying to have the "Selection is denoted by that item's RadioButton icon being checked" in both SelectedItem and SelectedIndex descriptions.

> | SelectedIndex | Gets or sets the selected item.|

Author: Ah! Okay will fix.

| HeaderTemplate | The template that specifies the visualization of the header object |
| SelectedItem | Gets or sets the selected item. Selection is denoted by that item's RadioButton icon being checked. |
| SelectedIndex | Gets or sets the selected index. Selection is denoted by that item's RadioButton icon being checked.|
| Items | Gets or sets an object source used to generate the content of the ItemsControl. All items placed within the RadioButtons control will get the content of a generated RadioButton. |
| ItemsSource | Gets or sets an object source used to generate the content of the ItemsControl. |
| ItemTemplate | The template set on any generated RadioButton elements within the RadioButtons control. If the root of the ItemTemplate is a RadioButton it will be used as the generated container. |
| MaxColumns | Defines the number of columns used to determine the control's layout |

## RadioButton Elements in the RadioButtons controls
Properties set on a RadioButton that is placed in a RadioButtons control are ignored by Narrator.

> For example, in this scenario:

Reviewer 3: Why do we have an example that's not in the examples section?

> For example, in this scenario:

Reviewer 3: Why do we have an example not in the Examples section?

```xaml

> <muxc:RadioButtons Header="App Mode">

Reviewer 3: To make this more realistic, each RadioButton should have its IsChecked property set to a different property (and be a Mode=TwoWay x:Bind)

>     <RadioButton GroupName="group test">test 1</RadioButton>

Reviewer 3: We don't need to demonstrate the useless case of setting GroupName when using RadioButtons.  I would take these out, but move the note to the Remarks section.

# Background

The RadioButtons control is our answer to enabling better keyboarding and narrator support for
RadioButton elements that need to be presented in a grouped layout.

Currently, the only way to group single RadioButton elements today is through using a StackPanel
and individually labeling each RadioButton through its GroupName property in order for narrator to
read out the options somewhat understandably.

However, using that method still leaves gaps in the narrator experience and StackPanel does not
behave the way users expect when navigating the list via keyboard - RadioButtons was introduced to
help solve those issues.

# Description

RadioButtons is a new control that enables you to create groups of RadioButton elements easily,
while also correctly supporting keyboarding and narrator functionality.

# Examples

The following example demonstrates how to create a small collective group of RadioButton elements
using the RadioButtons control.

```xaml
<muxc:RadioButtons Header="App Mode" SelectedIndex="2">
    <x:String>Dark</x:String>
    <x:String>Light</x:String>
    <x:String>Windows Default</x:String>         
</muxc:RadioButtons>
```

> ![alt text](DefaultRadioButtonGroup.png)

Reviewer 4: Can you update the image to show Dark/Light/Windows Default, instead of Item1/Item2/Items3? And actually show the selected index as selected.

## DataBinding with RadioButtons
An example of simple data binding of the ItemsSource with RadioButtons.

```xaml

> <preview:RadioButtons Header="App Mode" ItemsSource="{x:Bind radioButtonItems}" />

Reviewer 4: I assume the preview namespace is temporary, so...
```suggestion
<muxc:RadioButtons Header="App Mode" ItemsSource="{x:Bind radioButtonItems}" />
```

Author: hmm, I already made this change... not sure why it didn't show up - but yes! it should all be muxc.

```

```C#
public sealed partial class MainPage : Page
{
    public class OptionDataModel
    {
        public string Label;
        public override string ToString()
        {
            return Label;
        }
    }

    List<OptionDataModel> radioButtonItems;

    public MainPage()
    {
        this.InitializeComponent();

        radioButtonItems = new List<OptionDataModel>();

>         radioButtonItems.Add(new OptionDataModel() { label = "Item 1" });

Reviewer 4: I'm guessing you're trying to show the same app mode example as above, just using data binding. So perhaps use the dark/light strings from the previous example?

}
}
```

## RadioButtons Primitives

```
namespace Microsoft.UI.Xaml.Controls.Primitives
{
[webhosthidden]
unsealed runtimeclass RadioButtonsListView : Windows.UI.Xaml.Controls.ListView
{
    RadioButtonsListView();
}

[webhosthidden]
[default_interface]
unsealed runtimeclass RadioButtonsListViewItem : Windows.UI.Xaml.Controls.ListViewItem
{
    RadioButtonsListViewItem();
}
}
```

# API Notes

| Name | Description |
|:-:|:--|
| Header | Places a text label above the group of RadioButton elements and is exposed to the UIA tree via the DataContext property on the RadioButtons control's ContentPresenter. |
| HeaderTemplate | The template that specifies the visualization of the header object |
| SelectedItem | Gets or sets the index of the selected item. Selection is denoted by that item's RadioButton icon being checked. |
| SelectedIndex | Gets or sets the selected item.|
| Items | Gets or sets an object source used to generate the content of the ItemsControl. All items placed within the RadioButtons control will get the content of a generated RadioButton. |
| ItemsSource | Gets or sets an object source used to generate the content of the ItemsControl. |

> | ItemTemplate | The template set on any generated RadioButton elements within the RadioButtons control. If the root of the ItemTemplate is a RadioButton it will be used as the generated container. |

Reviewer 4: Perhaps an example of using ItemTemplate would be helpful?

Author: I believe I added this. Resolving.

> | MaxColumns | Defines the number of columns to divide the RadioButton group items into. |

Reviewer 4: ```suggestion
| MaxColumns | Defines the number of columns used to determine the control's layout |
```

## RadioButton Elements in the RadioButtons controls

> Properties set on a RadioButton that is placed in a RadioButtons control are ignored by the narrator.

Reviewer 4: ```suggestion
Properties set on a RadioButton that is placed in a RadioButtons control are ignored by Narrator.
```

For example, in this scenario:

```xaml
<muxc:RadioButtons Header="App Mode">

>     <RadioButton GroupName="group test">test 1</RadioButton>

Reviewer 4: It definitely wasn't clear to me from earlier in the spec that this scenario (adding RadioButton controls into RadioButtons control) was supported. Not sure how that can be clarified earlier.

Author: This should be in the examples above, *however* there appears to be a bug with it and I need @teaP to help address them before I add more examples to this spec.

Reviewer 2: This scenario is not supported. Is this a new requirement?

Author: It should be now with the migration from ListView to ItemsRepeater. Resolving this.
